Support app-defined ACLs via user-data table and whoAmI RPC

eddyashton opened the pull request:
This is primarily completing #431 with some minor additions. To get there, I also needed the caller lookup RPCs mentioned in #575, and to demonstrate this functionality I updated the `txregulator` sample app.

So 3 things:
- A `user_data` field in the `UserInfo` object, writable by members via consensus. This can store arbitrary json objects, so its up to the app instance to determine how permissions are encoded.
- New `whoAmI` and `whoIs` RPCs, so that an external user can find out their CCF User or Member ID.
- Changes to the `txregulator` sample app to check `user_data` permissions for registering regulators and banks. Most of the changes in the client are about drawing a clear distinction between the local name for a user (used as the `X` in `userX_privk.pem` for our Python infrastructure) and their CCF ID.

cimetrics commented:
![images](https://raw.githubusercontent.com/microsoft/CCF/cimetrics/cimetrics/image2019-11-26%2010%3A03%3A51.858734.png)

eddyashton commented on tests/infra/ccf.py:
```diff
@@ -172,8 +172,8 @@ def start_and_join(self, args):
 
-        self.consortium = infra.consortium.Consortium([1, 2, 3])
-        self.initial_users = [1, 2, 3]
+        self.consortium = infra.consortium.Consortium([0, 1, 2])
```
Previously we created a `member1_cert.pem` and `member1_privk.pem`, and these were associated with the member with ID `0` inside CCF. This changes the default member (and user) names used locally to match their eventual CCF IDs. This results in all of the `-1` changes in `memberclient.py`. In general we should probably try to break the assumption that these IDs are in-sync to avoid domino bugs - maybe the default users should be `alice`, `bob`, and `charlie`?

cimetrics commented:
![images](https://raw.githubusercontent.com/microsoft/CCF/cimetrics/cimetrics/image2019-11-26%2010%3A44%3A08.762884.png)

jumaffre commented on src/node/rpc/frontend.h:
```diff
@@ -434,6 +434,26 @@ namespace ccf
+        {
+          return jsonrpc::error(
+            jsonrpc::StandardErrorCodes::INVALID_PARAMS,
+            fmt::format("Cert not recognised"));
```
"Certificate not recognised"? Also no need for `fmt::format()`?

jumaffre commented on src/node/rpc/frontend.h:
```diff
@@ -434,6 +434,26 @@ namespace ccf
+      auto who_is = [this](Store::Tx& tx, const nlohmann::json& params) {
+        const WhoIs::In in = params;
+
+        auto certs_view = tx.get_view(*certs);
```
This will dereference a `nullptr` when executed on the nodes frontend. We should check that `certs != nullptr` here and return an error when it is.

eddyashton replied:
I've added nullptr guards here and below. It would be neater to just avoid installing these methods `if (certs == nullptr)`, but I don't want to introduce that pattern (complicating the "what methods are available" story) right now. At some point I'd like to enforce this distinction more cleanly, perhaps with a deeper inheritance hierarchy of `Frontend`s

jumaffre commented on src/node/rpc/frontend.h:
```diff
@@ -434,6 +434,26 @@ namespace ccf
         };
 
+      auto who_am_i = [this](const RequestArgs& args) {
+        return jsonrpc::success(WhoAmI::Out{args.caller_id});
```
On the nodes frontend, `args.caller_id` will be set to `INVALID_ID`. We should return an error if `certs == nullptr` instead.

jumaffre commented on src/node/rpc/calltypes.h:
```diff
@@ -68,6 +68,26 @@ namespace ccf
 
+  struct CallerInfo
+  {
+    size_t caller_id;
```
`CallerId caller_id`?

jumaffre commented on src/node/rpc/memberfrontend.h:
```diff
@@ -120,6 +120,21 @@ namespace ccf
          }},
+        {"set_user_data",
+         [this](Store::Tx& tx, const nlohmann::json& args) {
+           const UserId user_id = args["user_id"];
```
Should the JSON `args` be cast to a suitable `struct` instead of addressing the JSON fields directly?

jumaffre commented:
Should this new feature be briefly documented in https://microsoft.github.io/CCF/developers/index.html?

cimetrics commented:
![images](https://raw.githubusercontent.com/microsoft/CCF/cimetrics/cimetrics/image2019-11-26%2013%3A03%3A31.817342.png)

eddyashton commented:
> Should this new feature be briefly documented in https://microsoft.github.io/CCF/developers/index.html?

I've added this to the member docs - a section on "Adding users" under "Common governance operations". I don't think there's a clean spot to put it in the developers documentation at the moment. If it proves generally useful, we can add a few example uses in time.

cimetrics commented:
![images](https://raw.githubusercontent.com/microsoft/CCF/cimetrics/cimetrics/image2019-11-26%2013%3A56%3A46.271435.png)

jumaffre commented:
> > Should this new feature be briefly documented in https://microsoft.github.io/CCF/developers/index.html?
> 
> I've added this to the member docs - a section on "Adding users" under "Common governance operations". I don't think there's a clean spot to put it in the developers documentation at the moment. If it proves generally useful, we can add a few example uses in time.

There's already a "Adding Users" section described when the network is opened (https://microsoft.github.io/CCF/members/open_network.html#adding-users). Should this two be merged somehow?

eddyashton commented:
> > > Should this new feature be briefly documented in https://microsoft.github.io/CCF/developers/index.html?
> > 
> > 
> > I've added this to the member docs - a section on "Adding users" under "Common governance operations". I don't think there's a clean spot to put it in the developers documentation at the moment. If it proves generally useful, we can add a few example uses in time.
> 
> There's already a "Adding Users" section described when the network is opened (https://microsoft.github.io/CCF/members/open_network.html#adding-users). Should this two be merged somehow?

Good spot! I've moved the new info under that existing section.

cimetrics commented:
![images](https://raw.githubusercontent.com/microsoft/CCF/cimetrics/cimetrics/image2019-11-26%2016%3A21%3A59.421127.png)
